contracts/src/Summa.sol

```diff
@@ -5,116 +5,159 @@ pragma solidity ^0.8.18;
+        cexAddressVerifier[addressType] = addressOwnershipVerifier;
+        emit AddressVerifierSet(addressType, _verifier);
+    }
+
     function submitProofOfAccountOwnership(
-        address[] memory _cexAddresses,
-        bytes[] memory cexSignatures,
-        string memory message
+        OwnedAddress[] memory _cexAddresses
     ) public {
-        require(
-            _cexAddresses.length == cexSignatures.length &&
-                _cexAddresses.length > 0,
-            "CEX addresses and signatures count mismatch"
-        );
-
         for (uint i = 0; i < _cexAddresses.length; i++) {
-            if (i >= cexAddresses.length) {
-                cexAddresses.push(_cexAddresses[i]);
-            } else if (_cexAddresses[i] != cexAddresses[i]) {
-                cexAddresses[i] = _cexAddresses[i];
-            }
-            address recoveredPubKey = keccak256(abi.encode(message))
-                .toEthSignedMessageHash()
-                .recover(cexSignatures[i]);
+            bytes32 addressHash = keccak256(_cexAddresses[i].cexAddress);
+            require(!confirmedAddress[addressHash], "Address already verified");
+            confirmedAddress[addressHash] = true;
             require(
-                _cexAddresses[i] == recoveredPubKey,
-                "Invalid signer for ETH address"
+                address(cexAddressVerifier[_cexAddresses[i].addressType]) !=
+                    address(0),
+                "Address verifier not set for this type of address"
+            );
+            require(
+                cexAddressVerifier[_cexAddresses[i].addressType]
+                    .verifyAddressOwnership(
+                        _cexAddresses[i].cexAddress,
+                        _cexAddresses[i].ownershipProof
+                    ),
+                "Invalid signer"
             );
         }
 
-        // Since we're always rewriting the old array with the new values, we need to make sure that we remove any leftovers if the new set of addresses is smaller than the old one
-        // TODO - explore some gas-efficient ways of maintaining this array
-        if (_cexAddresses.length < cexAddresses.length) {
-            for (uint i = _cexAddresses.length; i < cexAddresses.length; i++) {
-                cexAddresses.pop();
-            }
-        }
-
-        emit ExchangeAddressesSubmitted(cexAddresses);
+        emit ExchangeAddressesSubmitted(_cexAddresses);
     }
 
     /**
      * @dev Submit proof of solvency for a CEX
-     * @param erc20ContractAddresses The addresses of the ERC20 token contracts that the CEX holds (e.g., USDT, USDC, DAI)
-     * @param assetSums The total asset sums to prove. ETH balance should be the first element, followed by ERC20 balances in the order of erc20ContractAddresses
+     * @param assets Assets owned by CEX
      * @param mstRoot The root of the Merkle sum tree
      * @param proof The ZK proof
+     * @param timestamp The timestamp at which the proof was generated
```

Comment:
More precisely, `timestamp` refers to the moment in time in which the balances of the users are snapshotted in order to generate the merkle sum tree. We can say that this is the timestap at which the liabilities of the CEX are snapshotted.